docs/refman/ppr-indexppds.8.pod
#
# mouse:~ppr/src/docs/refman/ppr-indexppds.8.pod
# Copyright 1995--2003, Trinity College Computing Center.
# Written by David Chappell.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# * Redistributions of source code must retain the above copyright notice,
# this list of conditions and the following disclaimer.
# 
# * Redistributions in binary form must reproduce the above copyright
# notice, this list of conditions and the following disclaimer in the
# documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE 
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
# POSSIBILITY OF SUCH DAMAGE.
#
# Last modified 15 January 2003.
#

=head1 NAME

B<ppr-indexppds> - Scan font directories and create PPR's font index

=head1 SYNOPSIS

B<ppr-indexppds>

=head1 DESCRIPTION

This program scans the directories listed in the [ppds] section of
F</etc/ppr/ppr.conf> and builds an index of the PPD files it finds in those
directories.  The web interface refers to this index when displaying the
list of available PPD files.

=head1 FILES

The PPD file database is stored in the file F</var/spool/ppr/ppdindex.db>.

=head1 SEE ALSO

"PPR, a PostScript Print Spooler"

=head1 HISTORY

PPR was written at Trinity College during 1993--2003.

=head1 AUTHOR

David Chappell, Trinity College Computing Center, Hartford, Connecticut.

=cut


History.txt
mouse:~ppr/src/History.txt (symbolic link)

This file contains a history of changes to PPR which occured after the
release of version 1.00.


1.00, 26 April 1995

First public release.


1.10, 27 April 1995--7 June 1995

Removed "From:" line from message passed to Sendmail by pprd when sending
a message listing the remaining printer faults.  This is so that Sendmail
will take the "real name" from the password file and put it in the "From:"
line.

Removed dotmatrix/input.epson, a GIF file which had been inadvertently
included in the distribution.

Moved the contents of include/atalk_streams.h into
interfaces/atalk_streams.c since it was not long enough to justify an
include file, nor was it of interest to any other program.

Added code to ppad to classify members of the group "ppradmin" as PPR
addministrators who can edit printer and group configurations.

Added -v switch to ppad.

Removed uid's 1-99 from the list of priveledged users.

Fixed ppad(8) so that it considers the group "ppradmin" as priveledged
rather than "ppr".

Changed ppr(1) so that it will accept the members of the group "pprprox" as
priveledged as well as the members of "ppr".  This is for security reasons
as one may not want to grant a user who has the privedge of using the -f
switch the priveledge of deleting other people's jobs.

Modified ppop(1) and pprd(8) so that user's can request to see the queue
entry for a single job.  Previously, the only options were all jobs for a
certain destination or all jobs in all queues.

Rewrote and renamed the undocumented "ppop qdebug" command to become the new
documented command "ppop details".  (This was done primarily to support
wwwppop, an experimental managment interface for PPR.  Since it was still
incomplete it was not released with this version of PPR.)

Renamed interfaces/atalk_streams.c to atalk_ali.c and
papsrv/papsrv_streams.c to papsrv_ali.c

Removed code from "make dirs" so that it doesn't create the directory
"/etc/ppr/groups.ppd".

Added sleep=N and eoftimeout=N options to tcpip interface.

Added -C _filename_ switch to ppr invokation in xppr.

Added "-default okay" to xmessage invokation in xppr.

Added the "ppop lpq" sub-command to the "ppop help" information, thus
making it a documented sub-command.

Extended automatic bin selection to multiple media types when "%%PageMedia:"
comments are present.  This has not yet been well tested.

Made "%%EndProcset" an acceptable substitute for "%%EndProcSet" due to a
reputed bug in QuickDraw GX.

Changed rule in ppr_infile.c for detecting PCL code so that documents with
only one escape would still be detected.


1.20, 8 June 1995--10 January 1996

Fixed code in ppr_infile.c so that the ppr -o switch will work even if there
is no "DefFiltOpts:" line in the printer or group configuration file.

Added '(' to list of likely second characters for HP escape codes
in ppr_infile.c

Replaced printing of errno value with strerror(errno) output in some places,
eliminated printing of errno=x from other places where both were printed.

Fixed ppr(1) so that it sets the environment variables "PATH", "IFS", and
"SHELL" so that filters and responders it invokes are less easy to subvert.

Removed setting of PATH from ~ppr/src/misc_filters/tex.perl and changed
shell name from "/usr/bin/tainperl" to "/usr/bin/perl".

Changed code which determins default responder address so that the
environment variables "USER" and "LOGNAME" are used in preference to the
current uid.

Added prompts in Configure to set the directory into which the links to PPR
binaries will be written and the path to the X-Windows binaries such as
xterm.

Modified papsrv so that it will pass the network and node number of the
client to ppr with the -r switch.  (This is for use with an experimental
AppleTalk responder.)

Modified papsrv.c, renaming cleanup() to termination_handler() and causing
it to call fatal() only the first time it is called.  Without this, name
could be left on the network if the CAP version received two terminate
signals in rapid sucession.

Added code to re-line-wrap responder messages for certain responders.
(The library function responder_width() returns the width for a responder.
Currently it is hard coded to return zero for all responders except "samba".
A return value of zero indicates that the message should not be
re-line-wrapped.)

Dropt reading of LPDEST and PRINTER by ppr(1).  (I can not really remember
why I put this feature in.  I don't think the reasons where very good.  It
may have been that I thought that the same printer would be available thru
two spooling systems so one might want to set the default destination for
both with one variable.  That now seems a little far fetched.)

Separated Solaris and Starserver sections of sysdep.h

Added PNG (Portable Network Graphics) format with magic number recognition
to ppr_infile.c.  This remains untested since I do not have any programs
which manipulate this graphics format.

In pprd.c, put sleep(1) between execl("pprdrv") and _exit(242) so that pprd
has time to record the PID before the child exits.  (If this is not done
and execl() fails immediately, a child which should have become pprdrv may
exit before the parent records its PID in the information structure for that
queue.  This result is that it will not recoginize the purpose of the child
when it receives notification of its death and will leave the queue in the
printing state indefinitely.)

Did same in pprdrv.c in code which executes interface.  Also changed
_exit(255) to _exit(242).  (Exit 242 is the new standard since PERL uses 255
as the exit code when die() is executed.)

Added automatic creation of "DefFiltOpts:" lines to ppad(8).  Manual
creation is supported by the the "ppad deffiltopts" and "ppad group
deffiltopts" command.

Modified ppop so that interactive mode restores the terminal settings when
it exits.  Also changed terminal setting code to use the POSIX method of

ppd/ppr-indexppds.c
/*
** mouse:~ppr/src/ppd/ppr-indexppds.c
** Copyright 1995--2002, Trinity College Computing Center.
** Written by David Chappell.
**
** Redistribution and use in source and binary forms, with or without
** modification, are permitted provided that the following conditions are met:
**
** * Redistributions of source code must retain the above copyright notice,
** this list of conditions and the following disclaimer.
** 
** * Redistributions in binary form must reproduce the above copyright
** notice, this list of conditions and the following disclaimer in the
** documentation and/or other materials provided with the distribution.
**
** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
** AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
** IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
** ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE 
** LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
** CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
** SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
** INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
** CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
** ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
** POSSIBILITY OF SUCH DAMAGE.
**
** Last modified 14 August 2002.
*/

#include "before_system.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <dirent.h>
#include <errno.h>
#include <ctype.h>
#ifdef INTERNATIONAL
#include <locale.h>
#include <libintl.h>
#endif
#include "gu.h"
#include "global_defines.h"
#include "util_exits.h"
#include "version.h"

const char myname[] = "ppr-indexppds";
const char ppr_conf[] = PPR_CONF;
const char section_name[] = "ppds";
const char ppdindex_db[] = VAR_SPOOL_PPR"/ppdindex.db";

/*==========================================================================
** This routine is called for each file found in the PPD
** directories.
==========================================================================*/
static int do_file(FILE *indexfile, const char filename[])
    {
    int ret;
    char *pline, *p;
    char *Manufacturer = NULL;
    char *Product = NULL;
    char *ModelName = NULL;
    char *NickName = NULL;
    char *ShortNickName = NULL;

    printf("  %s", filename);

    if((ret = ppd_open(filename, stderr)) != EXIT_OK)
	return ret;

    while((pline = ppd_readline()))
	{
	if(!Manufacturer && (p = lmatchp(pline, "*Manufacturer:")) && *p == '"')
	    {
	    p++;
	    Manufacturer = gu_strndup(p, strcspn(p, "\""));
	    continue;
	    }
	if(!Product && (p = lmatchp(pline, "*Product:")) && *p == '"' && p[1] == '(')
	    {
	    p += 2;
	    Product = gu_strndup(p, strcspn(p, ")"));
	    continue;
	    }
	if(!ModelName && (p = lmatchp(pline, "*ModelName:")) && *p == '"')
	    {
	    p++;
	    ModelName = gu_strndup(p, strcspn(p, "\""));
	    continue;
	    }
	if(!NickName && (p = lmatchp(pline, "*NickName:")))
	    {
	    p++;
	    NickName = gu_strndup(p, strcspn(p, "\""));
	    continue;
	    }
	/* ShortNickName is only valid if it comes before NickName. */
	if(!ShortNickName && !NickName && (p = lmatchp(pline, "*ShortNickName:")))
	    {
	    p++;
	    ShortNickName = gu_strndup(p, strcspn(p, "\""));
	    continue;
	    }
	}

    #ifdef DEBUG
    printf(": Manufacturer=\"%s\", Product=\"%s\", ModelName=\"%s\", NickName=\"%s\", ShortNickName=\"%s\"",
	Manufacturer ? Manufacturer : "",
	Product ? Product : "",
	ModelName ? ModelName : "",
	NickName ? NickName : "",
	ShortNickName ? ShortNickName : ""
	);
    #endif

    fputc('\n', stdout);

    {
    char *vendor;
    const char *description;

    if(Manufacturer)
	{
	vendor = Manufacturer;
	}
    else if(NickName && isupper(NickName[0]) && isalpha(NickName[1]) && strchr(NickName, ' '))
	{
	vendor = gu_strndup(NickName, strcspn(NickName, " -"));
	}
    else
	{
	vendor = NULL;
	}

    if(ShortNickName)
	description = ShortNickName;
    else if(NickName)
    	description = NickName;
    else if(ModelName)
    	description = ModelName;
    else if(Product)
    	description = Product;
    else
    	description = filename;

    fprintf(indexfile, "%s:%s:%s\n",
	filename,
	vendor ? vendor : "Other",
	description
	);

    if(vendor && vendor != Manufacturer)
    	gu_free(vendor);
    }

    if(Manufacturer)
    	gu_free(Manufacturer);
    if(Product)
    	gu_free(Product);
    if(ModelName)
    	gu_free(ModelName);
    if(NickName)
    	gu_free(NickName);
    if(ShortNickName)
    	gu_free(ShortNickName);

    return EXIT_OK;
    } /* end of do_file() */

/*============================================================================
** This routine is called for each PPD directory.
============================================================================*/
static int do_dir(FILE *indexfile, const char dirname[])
    {
    DIR *dirobj;
    struct dirent *fileobj;
    char filename[MAX_PPR_PATH];
    struct stat statbuf;
    int retval = EXIT_OK;

    printf("%s\n", dirname);

    if(!(dirobj = opendir(dirname)))
    	{
    	fprintf(stderr, "%s: diropen(\"%s\") failed, errno=%d (%s)\n", myname, dirname, errno, gu_strerror(errno));
    	return EXIT_INTERNAL;
    	}

    while((fileobj = readdir(dirobj)))
	{
	/* Skip hidden files and the "." and ".." directories. */
	if(fileobj->d_name[0] == '.') continue;

	/* Build the full name of the file or directory. */
	ppr_fnamef(filename, "%s/%s", dirname, fileobj->d_name);

	if(stat(filename, &statbuf) == -1)
	    {
	    fprintf(stderr, "%s: stat() failed on \"%s\", errno=%d (%s)\n", myname, filename, errno, gu_strerror(errno));
	    return EXIT_INTERNAL;
	    }

	/* Process regular files as PPD files. */
	if(S_ISREG(statbuf.st_mode))
	    {
            if((retval = do_file(indexfile, filename)) != EXIT_OK)
                break;
	    continue;
	    }
	}

    closedir(dirobj);

    return retval;
    } /* end of do_dir() */

/*
** Command line options:
*/
static const char *option_chars = "";
static const struct gu_getopt_opt option_words[] =
	{
	{"help", 1000, FALSE},
	{"version", 1001, FALSE},
	{(char*)NULL, 0, FALSE}
	} ;

/*
** Print help.
*/
static void help_switches(FILE *outfile)
    {
    fputs(_("Valid switches:\n"), outfile);

    fputs(_(	"\t--version\n"
		"\t--help\n"), outfile);
    }

int main(int argc, char *argv[])
    {
    FILE *indexfile;
    FILE *cf;
    struct GU_INI_ENTRY *section;
    int i;
    const char *dirname;
    int retval = EXIT_OK;

    /* Initialize international messages library. */
    #ifdef INTERNATIONAL
    setlocale(LC_MESSAGES, "");
    bindtextdomain(PACKAGE, LOCALEDIR);
    textdomain(PACKAGE);
    #endif

    /* Parse the options. */
    {
    struct gu_getopt_state getopt_state;
    int optchar;
    gu_getopt_init(&getopt_state, argc, argv, option_chars, option_words);
    while((optchar = ppr_getopt(&getopt_state)) != -1)
    	{
    	switch(optchar)
    	    {
	    case 1000:			/* --help */
	    	help_switches(stdout);
	    	exit(EXIT_OK);

	    case 1001:			/* --version */
		puts(VERSION);
		puts(COPYRIGHT);
		puts(AUTHOR);
	    	exit(EXIT_OK);

	    default:			/* other getopt errors or missing case */
		gu_getopt_default(myname, optchar, &getopt_state, stderr);
	    	exit(EXIT_SYNTAX);
		break;
    	    }
    	}
    }

    /* open ppr.conf */
    if(!(cf = fopen(ppr_conf, "r")))
    	{
	fprintf(stderr, _("%s: can't open \"%s\", errno=%d (%s)\n"), myname, ppr_conf, errno, gu_strerror(errno));
	return EXIT_INTERNAL;
    	}

    /* Fetch the [ppds] section */
    if(!(section = gu_ini_section_load(cf, section_name)))
    	{
    	fprintf(stderr, _("%s: no [%s] section in \"%s\"\n"), myname, section_name, ppr_conf);
    	return EXIT_INTERNAL;
    	}

    fclose(cf);

    /* Create the output file */
    if(!(indexfile = fopen(ppdindex_db, "w")))
    	{
	fprintf(stderr, _("%s: can't create \"%s\", errno=%d (%s)\n"), myname, ppdindex_db, errno, gu_strerror(errno));
	return EXIT_INTERNAL;
    	}

    /* Warn */
    fprintf(indexfile, "# This file format is a temporary hack.  Its format will change.\n");

    /* iterate over the nameless values */
    for(i=0; section[i].name; i++)
	{
	if(section[i].name[0] == '\0')
	    {
	    dirname = gu_ini_value_index(&section[i], 0, "<MISSING VALUE>");
	    if((retval = do_dir(indexfile, dirname)) != EXIT_OK)
	   	break;
	    }
	}

    /* Free the in-memory representation of the [PPDs] section. */
    gu_ini_section_free(section);

    /* Close the completed PPD index file. */
    fclose(indexfile);

    return retval;

    } /* end of main() */

/* end of file */
